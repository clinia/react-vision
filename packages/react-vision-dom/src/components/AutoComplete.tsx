import React, { Component } from 'react';
import {
  escapeRegExp,
  createClassNames,
  extractInputEventsFromProps,
} from '../core/utils';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import {
  defaultLoadingIndicator,
  defaultClear,
  defaultSubmit,
} from './defaultComponents';
import { translatable } from 'react-vision-core';

type Suggestion = {
  id: string;
  suggestion: string;
};

interface Props {
  currentRefinement?: string;
  className?: string;
  refine?: (value: string | null) => void;
  searchForSuggestions: (value: string) => void;
  suggestions?: Array<Suggestion>;
  style?: React.CSSProperties;
  translate?: any;

  loadingIndicator?: React.ReactNode;
  clear?: React.ReactNode;
  submit?: React.ReactNode;
  renderSuggestion?: (suggestion: Suggestion) => React.ReactNode;

  autoFocus?: boolean;

  onSubmit?: (event: any) => void;
  onClear?: (event: any) => void;
  onSuggestionSelected?: (suggestion: Suggestion) => void;
  onBlur(event: React.FocusEvent<HTMLInputElement>);
  onFocus(event: React.FocusEvent<HTMLInputElement>);
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;

  isSearchStalled: boolean;
  showLoadingIndicator?: boolean;
  disabled?: boolean;
}

interface DefaultProps {
  currentRefinement: string;
  suggestions: Array<Suggestion>;
  className: string;
  searchForSuggestions: (value: string) => void;
  refine: (value: string | null) => void;
  autoFocus: boolean;
  showLoadingIndicator: boolean;
  isSearchStalled: boolean;
  loadingIndicator: React.ReactNode;
  clear: React.ReactNode;
  submit: React.ReactNode;
  renderSuggestion: (suggestion: Suggestion) => React.ReactNode;
  __inputRef: (el: any) => void;
}

type PropsWithDefaults = Props & DefaultProps;

type State = {
  query: string;
  isInputFocused: boolean;
};

const cx = createClassNames('AutoComplete');

class AutoComplete extends Component<PropsWithDefaults, State> {
  input!: HTMLInputElement;

  static propTypes = {
    currentRefinement: PropTypes.string,
    className: PropTypes.string,
    refine: PropTypes.func.isRequired,
    searchForSuggestions: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,

    loadingIndicator: PropTypes.node,
    clear: PropTypes.node,
    submit: PropTypes.node,

    autoFocus: PropTypes.bool,

    onSubmit: PropTypes.func,
    onClear: PropTypes.func,
    onChange: PropTypes.func,
    onBlur: PropTypes.func,
    onFocus: PropTypes.func,
    onSuggestionSelected: PropTypes.func,

    renderSuggestion: PropTypes.func,
    isSearchStalled: PropTypes.bool,
    showLoadingIndicator: PropTypes.bool,
    disabled: PropTypes.bool,
    __inputRef: PropTypes.func,
  };

  static defaultProps = {
    currentRefinement: '',
    className: '',
    style: {},
    suggestions: [],

    loadingIndicator: defaultLoadingIndicator(cx),
    clear: defaultClear(cx),
    submit: defaultSubmit(cx),

    autoFocus: false,
    isSearchStalled: false,
    showLoadingIndicator: false,
    disabled: false,
  };

  constructor(props: PropsWithDefaults) {
    super(props);
    this.state = {
      query: '',
      isInputFocused: false,
    };

    // We bind functions for test purposes instead of using arrow functions
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onClear = this.onClear.bind(this);
    this.onBlur = this.onBlur.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onSuggestionSelected = this.onSuggestionSelected.bind(this);
  }

  onBlur(event: React.FocusEvent<HTMLInputElement>) {
    const { onBlur } = this.props;

    this.setState({ isInputFocused: false });

    if (onBlur) {
      onBlur(event);
    }
  }

  onFocus(event: React.FocusEvent<HTMLInputElement>) {
    const { onFocus } = this.props;

    this.setState({ isInputFocused: true });

    if (onFocus) {
      onFocus(event);
    }
  }

  onSuggestionSelected(suggestion: Suggestion) {
    const { onSuggestionSelected } = this.props;

    console.log(suggestion);

    this.setState({ query: suggestion.suggestion });

    if (onSuggestionSelected) {
      onSuggestionSelected(suggestion);
    }
  }

  onChange(event: React.ChangeEvent<HTMLInputElement>) {
    const { searchForSuggestions, onChange } = this.props;
    const query = event.target.value;

    this.setState({ query });

    searchForSuggestions(query);

    if (onChange) {
      onChange(event);
    }
  }

  onSubmit(event: React.FormEvent<HTMLFormElement>) {
    const { refine, currentRefinement } = this.props;

    event.preventDefault();
    event.stopPropagation();
    this.input.blur();

    refine(currentRefinement);
  }

  onClear(event: React.FormEvent<HTMLFormElement>) {
    const { refine, onClear } = this.props;
    refine('');

    this.setState({ query: '' });
    if (onClear) {
      onClear(event);
    }
  }

  renderSuggestion = suggestion => {
    const { query } = this.state;

    const highlightedSuggestion = suggestion.suggestion.replace(
      new RegExp(escapeRegExp(query), 'gi'),
      match => {
        return `<strong>${match}</strong>`;
      }
    );

    return <div dangerouslySetInnerHTML={{ __html: highlightedSuggestion }} />;
  };

  onInputMount = input => {
    this.input = input;
    if (this.props.__inputRef) {
      this.props.__inputRef(input);
    }
  };

  render() {
    const {
      suggestions,
      translate,
      className,
      style,
      isSearchStalled,
      submit,
      clear,
      loadingIndicator,
      autoFocus,
      disabled,
      renderSuggestion,
    } = this.props;

    const { query, isInputFocused } = this.state;

    //Events that cannot be completely overridden due to internal use
    const internalEvents = [
      'onsubmit',
      'onclear',
      'onchange',
      'onblur',
      'onfocus',
      'onsuggestionselected',
    ];

    const autoCompleteInputEvents = extractInputEventsFromProps(
      internalEvents,
      this.props
    );

    return (
      <div className={classnames(cx(''), className)} style={style}>
        <form
          noValidate
          onSubmit={this.props.onSubmit ? this.props.onSubmit : this.onSubmit}
          onReset={this.onClear}
          className={cx('form', isSearchStalled ? 'form--stalledSearch' : '')}
          action=""
          role="search"
        >
          <div>
            <input
              ref={this.onInputMount}
              type="search"
              placeholder={translate('placeholder')}
              autoFocus={autoFocus}
              disabled={disabled}
              autoComplete="off"
              autoCorrect="off"
              autoCapitalize="off"
              spellCheck={false}
              required
              maxLength={512}
              value={query!}
              onChange={this.onChange}
              onBlur={this.onBlur}
              onFocus={this.onFocus}
              {...autoCompleteInputEvents}
              className={cx('input')}
            />
            <ul>
              {isInputFocused &&
                suggestions.map(suggestion => (
                  <li
                    key={suggestion.suggestion}
                    onMouseDown={() => this.onSuggestionSelected(suggestion)}
                  >
                    {renderSuggestion
                      ? renderSuggestion(suggestion)
                      : this.renderSuggestion(suggestion)}
                  </li>
                ))}
            </ul>
          </div>
          <button type="submit" title={translate('searchTitle')}>
            {submit}
          </button>
          <button
            type="reset"
            title={translate('clearTitle')}
            className={cx('clear')}
            hidden={!query || isSearchStalled}
          >
            {clear}
          </button>
          {this.props.showLoadingIndicator && (
            <span hidden={!isSearchStalled} className={cx('loadingIndicator')}>
              {loadingIndicator}
            </span>
          )}
        </form>
      </div>
    );
  }
}

export default translatable({
  placeholder: 'Default',
  searchTitle: 'Search',
  clearTitle: 'Clear',
})(AutoComplete);
