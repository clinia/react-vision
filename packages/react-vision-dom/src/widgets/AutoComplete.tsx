import { connectAutoComplete } from 'react-vision-core';
import AutoComplete from '../components/AutoComplete';

/**
 * @module AutoComplete
 */

/**
 * The AutoComplete component displays a search box that lets the user search for a specific query.
 * @alias module:AutoComplete
 * @kind widget
 * @prop {boolean} autoFocus=false - Set autoFocus to the autocomplete input.
 * @prop {string} className - Add a custom CSS class to the AutoComplete form container.
 * @prop {node} clear - Change the apparence of the default reset button (cross).
 * @prop {string} clearTitle - The reset button title.
 * @prop {string} defaultRefinement - Provide default refinement value when component is mounted.
 * @prop {string} highlightPostTag - The post tag that will highlight the matched part of the query in each suggestion. Usually a html tag like `</strong>`.
 * @prop {string} highlightPreTag - The pre tag that will highlight the matched part of the query in each suggestion. Usually a html tag like `<strong>`.
 * @prop {node} loadingIndicator - Change the apparence of the default loading indicator (spinning circle).
 * @prop {function} onClear - Listen to `reset` event sent from the AutoComplete form container.
 * @prop {function} onSubmit - Intercept submit event sent from the AutoComplete form container.
 * @prop {function} onSuggestionSelected - Executes every time that a suggestion is selected.
 * @prop {function} on* - Listen to any events sent from the search input itself.
 * @prop {string} placeholder - The label of the input placeholder.
 * @prop {function} renderSuggestion - Define how suggestions will be rendered.
 * @prop {string} searchTitle - The submit button title.
 * @prop {number} size=5 - Number of suggestions to show.
 * @prop {boolean} showLoadingIndicator=false - Display that the search is loading. This only happens after a certain amount of time to avoid a blinking effect. This timer can be configured with `stalledSearchDelay` props on <InstantSearch>. By default, the value is 200ms.
 * @prop {node} submit - Change the apparence of the default submit button (magnifying glass).
 * @prop {object} style - Add a custom React.CSSProperties object to the AutoComplete form container.
 * @prop {boolean} triggerSubmitOnSuggestionSelected=false - Define if the AutoComplete form container should be submitted onSuggestionSelected.
 * @themekey cvi-AutoComplete - The root div of the widget.
 * @themekey cvi-AutoComplete-clear - The reset button used to clear the content of the input.
 * @themekey cvi-AutoComplete-clearIcon - The default reset icon used inside the reset button.
 * @themekey cvi-AutoComplete-form - The wrapping form.
 * @themekey cvi-AutoComplete-input - The search input.
 * @themekey cvi-AutoComplete-loadingIcon - The default loading icon.
 * @themekey cvi-AutoComplete-loadingIndicator - The loading indicator container.
 * @themekey cvi-AutoComplete-submit - The submit button.
 * @themekey cvi-AutoComplete-submitIcon - The default magnifier icon used with the search input.
 * @example
 * import React from 'react';
 * import cliniasearch from 'cliniasearch/lite';
 * import { Vision, AutoComplete } from 'react-vision-dom';
 *
 * const searchClient = cliniasearch(
 *   'latency',
 *   '6be0576ff61c053d5f9a3225e2a90f76'
 * );
 *
 * const App = () => (
 *   <Vision
 *     searchClient={searchClient}
 *   >
 *     <AutoComplete />
 *   </Vision>
 * );
 */
export default connectAutoComplete(AutoComplete)
