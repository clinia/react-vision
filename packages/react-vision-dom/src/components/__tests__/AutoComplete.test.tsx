import React from 'react';
import renderer from 'react-test-renderer';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AutoComplete from '../AutoComplete';

Enzyme.configure({ adapter: new Adapter() });

describe('SearchBox', () => {
  it('applies its default props', () => {
    const instance = renderer.create(<AutoComplete refine={() => null} />);

    expect(instance.toJSON()).toMatchSnapshot();

    instance.unmount();
  });

  it('applies its default props with custom className', () => {});
  it('applies its default props with custom style', () => {});
  it('transfers the autoFocus prop to the underlying input element', () => {});
  it('transfers the disabled prop to the underlying input element', () => {});
  it('lets you give custom components for clear and submit', () => {});
  it('lets you customize its translations', () => {});
  it('should render the loader if showLoadingIndicator is true', () => {});
  it('onSubmit behavior should be override if provided as props', () => {});
  it('should accept `on*` events', () => {});
});
