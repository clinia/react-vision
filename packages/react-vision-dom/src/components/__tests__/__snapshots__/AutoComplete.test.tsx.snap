// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox applies its default props 1`] = `
<div
  className="cvi-AutoComplete"
  style={Object {}}
>
  <form
    action=""
    className="cvi-AutoComplete-form cvi-AutoComplete"
    noValidate={true}
    onReset={[Function]}
    onSubmit={[Function]}
    role="search"
  >
    <div>
      <input
        autoCapitalize="off"
        autoComplete="off"
        autoCorrect="off"
        autoFocus={false}
        className="cvi-AutoComplete-input"
        disabled={false}
        maxLength={512}
        placeholder="Default"
        required={true}
        spellCheck={false}
        type="search"
        value=""
      />
      <ul />
    </div>
    <button
      title="Search"
      type="submit"
    />
    <button
      className="cvi-AutoComplete-clear"
      hidden={true}
      title="Clear"
      type="reset"
    />
  </form>
</div>
`;
